dipy/denoise/localpca.py

```diff
@@ -145,19 +209,42 @@ def localpca(arr, sigma, mask=None, pca_method='eig', patch_radius=2,
                     C = C / X.shape[0]
                     [d, W] = eigh(C, turbo=True)
 
+                if sigma is None:
+                    # Random matrix theory
+                    this_var, ncomps = _pca_classifier(d, patch_size ** 3)
+                else:
+                    # Predefined variance
+                    this_var = var[i, j, k]
+
                 # Threshold by tau:
-                W[:, d < tau] = 0
+                tau = tau_factor ** 2 * this_var
+
+                # Update ncomps according to tau_factor
+                ncomps = np.sum(d < tau)
+                W[:, :ncomps] = 0
+
                 # This is equations 1 and 2 in Manjon 2013:
                 Xest = X.dot(W).dot(W.T) + M
                 Xest = Xest.reshape(patch_size,
                                     patch_size,
-                                    patch_size, arr.shape[-1])
+                                    patch_size, dim)
                 # This is equation 3 in Manjon 2013:
-                this_theta = 1.0 / (1.0 + np.sum(d > 0))
+                this_theta = 1.0 / (1.0 + dim - ncomps)
                 theta[ix1:ix2, jx1:jx2, kx1:kx2] += this_theta
                 thetax[ix1:ix2, jx1:jx2, kx1:kx2] += Xest * this_theta
+                if return_sigma is True and sigma is None:
+                    var[ix1:ix2, jx1:jx2, kx1:kx2] += this_var * this_theta
+                    thetavar[ix1:ix2, jx1:jx2, kx1:kx2] += this_theta
 
     denoised_arr = thetax / theta
     denoised_arr.clip(min=0, out=denoised_arr)
     denoised_arr[~mask] = 0
```

Comment:
@RafaelNH this is not your fault, but just a previous bug I guess: 
Can you change this to..
`denoised_arr[mask==0] = 0`